torch/csrc/jit/codegen/cuda/root_domain_map.h

```diff
@@ -0,0 +1,159 @@
+#pragma once
+
+#include <torch/csrc/jit/codegen/cuda/iter_visitor.h>
+#include <torch/csrc/jit/codegen/cuda/utils.h>
+
+#include <torch/csrc/WindowsTorchApiMacro.h>
+
+namespace torch {
+namespace jit {
+namespace fuser {
+namespace cuda {
+
+class DomainKey {
+ public:
+  DomainKey() = default;
+  DomainKey(const TensorDomain* td, const IterDomain* id) : td_(td), id_(id) {}
+  const TensorDomain* td() const {
+    return td_;
+  }
+  const IterDomain* id() const {
+    return id_;
+  }
+  bool operator==(const DomainKey& other) const {
+    return td() == other.td() && id() == other.id();
+  }
+  std::ostream& print(std::ostream& os) const;
+
+ private:
+  const TensorDomain* td_ = nullptr;
+  const IterDomain* id_ = nullptr;
+};
+
+inline std::ostream& operator<<(std::ostream& os, const DomainKey& key) {
+  return key.print(os);
+}
+
+struct DomainKeyHash {
+  std::size_t operator()(const DomainKey& key) const {
+    return std::hash<const TensorDomain*>{}(key.td()) ^ std::hash<const IterDomain*>{}(key.id());
+  }
+};
+
+class TORCH_CUDA_API FindIncompatibleDomains : private IterVisitor {
```

Comment:
Please comment. It seems what you're looking for is any consumer of a reduction cannot be mapped to any iteration domain before that reduction that was reduced. You're having a hard mapping break at reduction IterDomains.